splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,439 @@
+  override def setUp(): Unit = splaySetup()
+  override def tearDown(): Unit = splayTearDown()
+
+  //// Configuration.
+  private final val kSplayTreeSize = 8000
+  private final val kSplayTreeModifications = 80
+  private final val kSplayTreePayloadDepth = 5
+
+  private var splayTree: SplayTree[js.Dynamic] = null
+  private var splaySampleTimeStart = 0L
+
+  private def generatePayloadTree(depth: Int, tag: String): js.Dynamic = {
+    if (depth == 0) {
+      js.Dynamic.literal(
+        array = js.Array(0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
+        string = s"String for key $tag in leaf node"
+      )
+    } else {
+      js.Dynamic.literal(
+        left =  generatePayloadTree(depth - 1, tag),
+        right = generatePayloadTree(depth - 1, tag)
+      )
+    }
+  }
+
+  Random.setSeed(1L)
+  private def generateKey(): Int = {
+    // The benchmark framework guarantees that Math.random is
+    // deterministic; see base.js.
+    Random.nextInt()
+  }
+
+  private var splaySamples: Int = 0
+  private var splaySumOfSquaredPauses = 0L
+
+  def splayRMS(): Long =
+    Math.round(Math.sqrt(splaySumOfSquaredPauses / splaySamples) * 10000)
+
+  private def splayUpdateStats(time: Long) = {
+    val pause = time - splaySampleTimeStart
+    splaySampleTimeStart = time
+    splaySamples += 1
+    splaySumOfSquaredPauses += pause * pause
+  }
+
+  private def insertNewNode(): Int = {
+    // Insert new node with a unique key.
+    var key = 0
+    do {
+      key = generateKey()
+    } while (splayTree.find(key) != null)
+    val payload = generatePayloadTree(kSplayTreePayloadDepth, key.toString)
+    splayTree.insert(key, payload)
+    key
+  }
+
+  private def splaySetup(): Unit = {
+    splayTree = new SplayTree
+    splaySampleTimeStart = System.nanoTime()
+    for (i <- 0 until kSplayTreeSize) {
+      insertNewNode()
+      if ((i+1) % 20 == 19)
+        splayUpdateStats(System.nanoTime())
+    }
+  }
+
+  private def splayTearDown(): Unit = {
+    // Allow the garbage collector to reclaim the memory
+    // used by the splay tree no matter how we exit the
+    // tear down function.
+    val keys = splayTree.exportKeys
+    splayTree = null
+
+    splaySamples = 0
+    splaySumOfSquaredPauses = 0
+
+    // Verify that the splay tree has the right size.
+    val length = keys.length
+    if (length != kSplayTreeSize)
+      throw new Exception("Splay tree has wrong size")
+
+    // Verify that the splay tree has sorted, unique keys.
+    for (i <- 0 until length - 1) {
+      if (keys(i) >= keys(i + 1))
+        throw new Exception("Splay tree not sorted")
+    }
+  }
+
+  private def splayRun(): Unit = {
+    // Replace a few nodes in the splay tree.
+    for (i <- 0 until kSplayTreeModifications) {
+      val key = insertNewNode()
+      val greatest = splayTree.findGreatestLessThan(key)
+      if (greatest == null) splayTree.remove(key)
+      else splayTree.remove(greatest.key)
+    }
+    splayUpdateStats(System.nanoTime())
```

Comment:
Is this part of the benchmark loop? If yes, it will kill you because of the `Long`s.